File: apps/ios/Pebbles/Features/Profile/Views/GroupPebblesByMonth.swift
import Foundation

/// Groups pebbles by the first day of their calendar month, returning
/// `(monthStart, pebbles)` pairs ordered descending by month.
///
/// - Keys are the first instant of each month in the provided calendar.
/// - Within a group, input order is preserved — callers typically pass
///   pebbles already sorted descending by `happenedAt`.
/// - `calendar` is injectable to keep tests deterministic; production
///   callers should pass `Calendar.current`.
func groupPebblesByMonth(
    _ pebbles: [Pebble],
    calendar: Calendar
) -> [(key: Date, value: [Pebble])] {
    let buckets = Dictionary(grouping: pebbles) { pebble -> Date in
        let comps = calendar.dateComponents([.year, .month], from: pebble.happenedAt)
        return calendar.date(from: comps) ?? pebble.happenedAt
    }
    return buckets
        .map { (key: $0.key, value: $0.value) }
        .sorted { $0.key > $1.key }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/profile/GroupPebblesByMonth.kt
package app.pbbls.android.features.profile

import app.pbbls.android.features.path.models.Pebble
import java.time.YearMonth
import java.time.ZoneId

/**
 * Groups pebbles by calendar month in [zone], returning `(month, pebbles)`
 * pairs ordered descending by month — ports iOS `groupPebblesByMonth`.
 *
 * - Within a group, input order is preserved — callers typically pass pebbles
 *   already sorted descending by `happenedAt`.
 * - [zone] is injectable to keep tests deterministic; production callers pass
 *   `ZoneId.systemDefault()` (the iOS `Calendar.current` analog).
 */
fun groupPebblesByMonth(
    pebbles: List<Pebble>,
    zone: ZoneId,
): List<Pair<YearMonth, List<Pebble>>> {
    val buckets = LinkedHashMap<YearMonth, MutableList<Pebble>>()
    for (pebble in pebbles) {
        val month = YearMonth.from(pebble.happenedAt.atZoneSameInstant(zone))
        buckets.getOrPut(month) { mutableListOf() }.add(pebble)
    }
    return buckets.entries
        .map { it.key to it.value.toList() }
        .sortedByDescending { it.first }
}
